### PR TITLE
initial attempt to fix parallel execution problem

### PR DESCRIPTION
**Pull request by dsyzhu**
Am not confident in the fix. So just comment out the code I think causes the problem

### CODE REVIEW

**Comment by dsyzhu** [comment]
Hi @xythian ,

When you have time can you review the PR? Do you think it is a valid fix?

Thanks,
Daisy

**Comment by xythian** [comment]
Hm, interesting. It is likely this bug also affects other valid queries which should run more in parallel than they do.

That would fix that issue but it may have undesirable performance side-effects (without that code *every* step will run as separate tasks, whereas the intended behavior was to avoid passing off to another task when possible to avoid that overhead -- the engine treats every discrete computation as a step, so removing the merges entirely may not be ideal). Ideally we'd have a perf test to quantify the impact either way, because if it's fast enough then the simple solution is to just stop merging by removing the commented out code -- it's only worth doing the more advanced solution if it's necessary.

The more advanced solution would involve fixing this buggy step merge. That code is too aggressive at collapsing into steps because the *intended* behavior was to collapse when the next task *can't start*  *and* when the candidate is the only input to it.

It needs to take into account what can be started rather than just what the inputs are and avoid merging steps which can run in parallel into the same task while still merging sequential steps that are serial into the same task.

**Comment by dsyzhu** [comment]
@xythian ,

Thanks for the comment.

You mean removing the code may introduce other overhead. Right?

Let me see if I can find a better solution.

-Daisy

**Comment by dsyzhu** [comment]
@xythian ,

I have updated PR. Can you review it when you have time?

Thanks

**Comment by xythian** [comment]
@dsyzhu I'll try to take a look over the weekend.

**Comment by dsyzhu** [comment]
@xythian ,

Thanks

File: yqlplus_engine/src/main/java/com/yahoo/yqlplus/engine/internal/tasks/GraphPlanner.java [hunk 1]
```diff
@@ -114,7 +116,20 @@ public ForkTask plan(Step root) {
...
+                            if (entry.getKey() instanceof OperatorStep) {
+                                if (!((OperatorStep) entry.getKey()).getCompute().getOperator().equals(PhysicalOperator.END)) {
+                                    if (!(entry.getValue().todo.size() == 1)) {
+                                        found = true;
+                                        break;
+                                    }
+                                }
+                            }
+                        }
+                        modified = true && found;
```
Comment (xythian) [inline yqlplus_engine/src/main/java/com/yahoo/yqlplus/engine/internal/tasks/GraphPlanner.java]: This could be 'modified = found'

Comment (dsyzhu) [reply yqlplus_engine/src/main/java/com/yahoo/yqlplus/engine/internal/tasks/GraphPlanner.java]: Let me  check. Thank you! About the test failures which jdk did you use?

Comment (xythian) [reply yqlplus_engine/src/main/java/com/yahoo/yqlplus/engine/internal/tasks/GraphPlanner.java]: I was using JDK 11, which I suspect is why I was seeing failures.

Comment (dsyzhu) [reply yqlplus_engine/src/main/java/com/yahoo/yqlplus/engine/internal/tasks/GraphPlanner.java]: Merge after changing 'modified = found'. Thanks

**Pull request merged**